### ISSUE TITLE
Twitter: Browser notifications are unsupported

### ISSUE THREAD

**Issue by SenpaiSharp** [BrowserWorks/waterfox#893]
Since updating to the latest version, I can no longer enable browser push notifications on Twitter. I get this message when going to the Web notifications section under the user settings:

> Browser notifications are unsupported
> This  browser doesn't support web notifications.

I have tried changing the site specific useragent to a few of the latest, most common firefox ones to no avail.

**Comment by laniakea64** [comment]
If you go to `about:config` and set `dom.serviceWorkers.enabled` to `true`, does that fix it?

(I don't know if restarting Waterfox is required for that to take effect.)

**Comment by SenpaiSharp** [comment]
> 
> 
> If you go to `about:config` and set `dom.serviceWorkers.enabled` to `true`, does that fix it?
> 
> (I don't know if restarting Waterfox is required for that to take effect.)

Seems to fix it but I never had to enable this before and would prefer not to.

**Comment by laniakea64** [comment]
> I never had to enable this before and would prefer not to.

It was enabled by default before.  see https://github.com/MrAlex94/Waterfox/commit/521afd50114f7814ad7f0c013fd7c3dfad15d4e5

**Comment by MrAlex94** [comment]
Surprised Twitter needs service workers for Notifications.

Btw service workers were enabled previously but disabled them due to some
issues.

**Comment by SenpaiSharp** [comment]
Fair enough, I was unaware that it was enabled before. I am confused reading up on it, as some places are inferring that no notifications would work without it enabled but plenty of other places I tried worked fine.

**Comment by grahamperrin** [comment]
From <https://github.com/MrAlex94/Waterfox/issues/545#issue-321359296> (2018-05-08): 

> … Push notifications rely on service workers …

**Comment by SenpaiSharp** [comment]
> Surprised Twitter needs service workers for Notifications.

Best I can tell, there's a check to see if they're enabled before they'll allow notifications to be enabled but they don't actually seem to use that type of notification. I'm guessing this is some weird mix of legacy and newer code getting their wires crossed but that's just my speculation.

To test this, I re-enabled them and installed the Block Service Workers extension:

https://addons.mozilla.org/en-US/firefox/addon/block-service-workers/

The extension works seems to do its job. It blocks on visiting Youtube and it blocks this push notifications test, for example:

https://gauntface.github.io/simple-push-demo/

But on Twitter, with the service workers enabled, it lets me allow web notifications and they do work but at no time is anything ever blocked by the extension.

**Comment by grahamperrin** [comment]
I get this: 

![image](https://user-images.githubusercontent.com/192271/54492459-c6ee4180-48be-11e9-8620-6a81f5ece5ff.png)

– and this: 

![image](https://user-images.githubusercontent.com/192271/54492475-eb4a1e00-48be-11e9-8932-12abe11c31c2.png)

– but then I don't know when something might occur to trigger a notification, because I'm really not into _following_: 

![image](https://user-images.githubusercontent.com/192271/54492597-f487ba80-48bf-11e9-8dc3-ace34f8ed3bf.png)

@SenpaiSharp I temporarily followed you, would you like to do something that might trigger a notification here?

**Comment by grahamperrin** [comment]
@SenpaiSharp maybe put that thought on hold. Sorry. 

I forgot, for test purposes I un-registered my two service workers for Twitter. 

Neither worker has been regained (and I have no [Service Worker Detector](https://addons.mozilla.org/addon/service-worker-detector/) icon in my address bar):

![image](https://user-images.githubusercontent.com/192271/54492924-a1fbcd80-48c2-11e9-8949-079e900bb24a.png) 

I'll re-enable my [Redirector](https://addons.mozilla.org/addon/redirector/) redirect to mobile Twitter, this might cause one of the two workers to reappear sooner rather than later …

**Comment by SenpaiSharp** [comment]
> @SenpaiSharp I temporarily followed you, would you like to do something that might trigger a notification here?

I got a notification when you followed me, actually.

For reference this is what I get when dom.serviceWorkers.enabled is set to False.

![settings4](https://user-images.githubusercontent.com/36146858/54493192-5a197e00-48a3-11e9-8790-454599929131.png)

And when set to true.

![settings 5](https://user-images.githubusercontent.com/36146858/54493203-7e755a80-48a3-11e9-8750-956d0c9157e1.png)


My settings page looks different than yours, I'm guessing it's just a Windows vs Mac thing. 

![settings2](https://user-images.githubusercontent.com/36146858/54493105-900a3280-48a2-11e9-8bee-2ad6e6a658c1.png)

As you can see, nothing has been blocked:

![settings3](https://user-images.githubusercontent.com/36146858/54493107-91d3f600-48a2-11e9-8b47-3d14ef8fda6f.png)

As I said, I got a notification when you followed me. I didn't screencap it because I didn't realize it was you but it did work. My only conclusion is that Twitter has a check on the settings page for service workers, but ultimately doesn't actually use or need them for their implementation. Why they even check, I don't know. If there's something that could fool it without having to set dom.serviceWorkers.enabled to true, that would be nice but in fairness, this feels like something Twitter should fix on their end.

Edit: Just for posterity, here's a notification via twitter:

![settings6](https://user-images.githubusercontent.com/36146858/54493508-3b1ceb00-48a7-11e9-849d-b4abc1d0d5a4.png)

and here's the Blocker working as intended, in conjunction with no conflict:

![settings7](https://user-images.githubusercontent.com/36146858/54493510-46701680-48a7-11e9-92e1-04ee152e19b3.png)

**Comment by grahamperrin** [comment]
Thanks. 

Side note, 

> My settings page looks different than yours, I'm guessing it's just a Windows vs Mac thing.

I use [Breezemite](https://store.kde.org/p/1169286/) with KDE – on FreeBSD-CURRENT (I switched away from Apple). 

The major differences in my screenshots were probably from me using a narrow window.

**Comment by grahamperrin** [comment]
@SenpaiSharp please clarify: 

> ![image](https://user-images.githubusercontent.com/36146858/54493508-3b1ceb00-48a7-11e9-849d-b4abc1d0d5a4.png) 

When that notification appeared, did you (or did you not) have a worker for Twitter at `about:serviceworkers`?

**Comment by SenpaiSharp** [comment]
Okay, apparently I **do** have one registered.  Neither Service Worker Detector nor Block Service Workers appear to catch it.

**Comment by grahamperrin** [comment]
The Service Worker Detector icon appears only when a worker is active. 

Block Service Workers does not have a retrospective effect. I clarified the first bullet point at https://github.com/MrAlex94/Waterfox/issues/545#issuecomment-473608315

**Comment by SenpaiSharp** [comment]
I can unregister it and Block Service Workers won't catch it when I reenable notifications on Twitter.

Compare to this test: https://gauntface.github.io/simple-push-demo/



After unregistering

![after](https://user-images.githubusercontent.com/36146858/54496343-97423800-48c4-11e9-82ff-5872bed7e6b6.png)

Twitter will waltz through without problem. Whatever the difference is between the Twitter and everything else, I'm afraid I can't say.

**Comment by grahamperrin** [comment]
> unregister it and Block Service Workers won't catch it

Did you close all Twitter tabs before un-registering? 

I regained a `https://mobile.twitter.com/sw.js` worker. I can't tell exactly when or how it was regained, but my `serviceworker.txt` file was modified 18:36 yesterday (Sunday).

**Comment by SenpaiSharp** [comment]
Yes. I've gone as far as using a fresh profile on a separate install. Twitter somehow always evades Block Service Workers.

**Comment by SenpaiSharp** [comment]
I've done more testing and a lot more confirmation that BSW  is completely evaded by Twitter. I don't know why and as this is far out of the scope of Waterfox at this time, I'm going to close this. I have decided to just move on with service workers entirely disabled and will find another solution for my notification needs.

**Comment by grahamperrin** [comment]
Maybe relevant: my service worker for `mobile.twitter.com` is one of only two that has an endpoint, endpoints are in the `updates.push.services.mozilla.com` area.

**Comment by grahamperrin** [comment]
FYI, with Chromium: [mobile Twitter peculiarities with Block Service Workers · Issue #15 · clod81/block_service_workers](https://github.com/clod81/block_service_workers/issues/15)

**Comment by grahamperrin** [comment]
> … I forgot, for test purposes I un-registered my two service workers for Twitter.
> 
> Neither worker has been regained

After regaining both workers, I un-registered the one for mobile Twitter then found the remaining worker apparently working for push notifications. 

### Screenshots

A notification at launch time: 

![2019-03-20 02:26:10 push notification from Twitter](https://user-images.githubusercontent.com/192271/54726289-6f73fe00-4b6a-11e9-84e1-ba7a0e7c7a0f.png)

For the only Twitter worker in Firefox, no push endpoint: 

![2019-03-20 02:28:03 no push endpoint in Firefox](https://user-images.githubusercontent.com/192271/54726353-b7932080-4b6a-11e9-984b-c8897c149132.png)

Waterfox, post-launch:

![2019-03-20 02:29:08 Waterfox initial tab not yet loaded](https://user-images.githubusercontent.com/192271/54726415-f3c68100-4b6a-11e9-926b-a8a602f4482d.png)

For the only Twitter worker in Waterfox, a push endpoint: 

![2019-03-20 02:30:45 push endpoint for non-mobile Twitter, not mobile](https://user-images.githubusercontent.com/192271/54726428-017c0680-4b6b-11e9-80e8-81f02c296232.png)
